ENH: Add Linux aarch64 and ppc64le and macOS builds

matthewfeickert opened the pull request:
**Update to v1 recipe format**

* c.f. https://conda-forge.org/blog/2025/02/27/conda-forge-v1-recipe-support/
* Use
```
pixi exec conda-recipe-manager convert --output recipe/recipe.yaml recipe/
```
* Use Git repository and build from source with autotools.
   - Use tags to keep autotick bot support.
* Disable compiling a static library to conform to [CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
   - Wasn't caught in https://github.com/conda-forge/staged-recipes/pull/23671.
* Add rattler-build and pixi as conda tools.
* Bump build number.

**Enable macOS builds**
* Enable IPV6 constants on macOS which are required to compile.
   - c.f. https://krbdev.mit.edu/rt/Ticket/History.html?id=8349

**Add test program**
* Add test program to ensure library is working as expected. The [test program is copied from the Homebrew recipe for `libtirpc`](https://github.com/Homebrew/homebrew-core/blob/f99472a12a14f7512309ebedd2fd44072737ac34/Formula/lib/libtirpc.rb#L29-L40).

**Add `linux-aarch64`, `linux-ppc64le`, and `osx-arm64` builds**
Add multi-platform support by enabling builds for all Unix platforms on conda-forge.
* c.f. https://conda-forge.org/docs/maintainer/conda_forge_yml/#build-platform

**Rerender feedstock**

* Use
```
pixi exec conda-smithy rerender -c
```

Context: [#general > XDR / TI-RPC multi-platform libraries on conda-forge?](https://conda-forge.zulipchat.com/#narrow/channel/457337-general/topic/XDR.20.2F.20TI-RPC.20multi-platform.20libraries.20on.20conda-forge.3F/with/540941712)

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [N/A] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

matthewfeickert committed:
MNT: Update to v1 recipe format

* Update to v1 recipe.
* Use Git repository and build from source with autotools.
   - Use tags to keep autotick bot support.
* Add rattler-build and pixi as conda tools.
* Bump build number.

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

matthewfeickert committed:
ENH: Add Linux aarch64 and ppc64le and macOS arm64 builds

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,43 @@
+
+source:
+  url: https://downloads.sourceforge.net/libtirpc/libtirpc-${{ version }}.tar.bz2
+  sha256: b47d3ac19d3549e54a05d0019a6c400674da716123858cfdb6d3bdd70a66c702
```
```console
$ curl -sL https://downloads.sourceforge.net/libtirpc/libtirpc-1.3.7.tar.bz2 | openssl sha256
SHA2-256(stdin)= b47d3ac19d3549e54a05d0019a6c400674da716123858cfdb6d3bdd70a66c702
```

I'm not sure why soureforge is giving different digests on conda-forge vs. locally.

cphyc replied:
That's a really good question. Locally, I also get the same sha256 as you.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,96 @@
+
+source:
+  git: git://linux-nfs.org/~steved/libtirpc
+  rev: 5bd87617e4b3129e76e79847c95bdea4b1290d9f
```
So this works, and doesn't get hit with sourceforge instability, though I don't know how/if automatic rerenders when new tags would be released would work. :?

jaimergp replied:
Don't they have tags upstream? I can see this:

```bash
$ git clone git://linux-nfs.org/\~steved/libtirpc
$ cd libtirpc
$ git tag | cat
0.0.9-rc1
0.0.10
0.0.10-rc1
libtirpc-0-1-10
libtirpc-0-2-0
libtirpc-0-2-1
libtirpc-0-2-1-rc1
libtirpc-0-2-1-rc2
libtirpc-0-2-1-rc3
libtirpc-0-2-2
libtirpc-0-2-2-rc1
libtirpc-0-2-2-rc2
libtirpc-0-2-2-rc3
libtirpc-0-2-2-rc4
libtirpc-0-2-3
libtirpc-0-2-3-rc1
libtirpc-0-2-3-rc2
libtirpc-0-2-3-rc3
libtirpc-0-2-3-rc4
libtirpc-0-2-4
libtirpc-0-2-4-rc1
libtirpc-0-2-4-rc2
libtirpc-0-2-4-rc3
libtirpc-0-2-5
libtirpc-0-2-5-rc1
libtirpc-0-2-5-rc2
libtirpc-0-2-5-rc3
libtirpc-0-2-5-rc4
libtirpc-0-2-6-rc1
libtirpc-0-2-6-rc2
libtirpc-0-2-6-rc3
libtirpc-0-2-6-rc4
libtirpc-0-2-6-rc5
libtirpc-0-3-0
libtirpc-0-3-1
libtirpc-0-3-1-rc1
libtirpc-0-3-1-rc2
libtirpc-0-3-2
libtirpc-0-3-3-rc1
libtirpc-0-3-3-rc2
libtirpc-0-3-3-rc3
libtirpc-1-0-1
libtirpc-1-0-2
libtirpc-1-0-2-rc1
libtirpc-1-0-2-rc2
libtirpc-1-0-2-rc3
libtirpc-1-0-2-rc4
libtirpc-1-0-3
libtirpc-1-0-3-rc1
libtirpc-1-0-3-rc2
libtirpc-1-0-3-rc3
libtirpc-1-0-4-rc1
libtirpc-1-0-4-rc2
libtirpc-1-1-4
libtirpc-1-1-5-rc1
libtirpc-1-1-5-rc2
libtirpc-1-1-5-rc3
libtirpc-1-1-5-rc4
libtirpc-1-2-5
libtirpc-1-2-6
libtirpc-1-2-6-rc1
libtirpc-1-2-6-rc2
libtirpc-1-2-7-rc1
libtirpc-1-2-7-rc2
libtirpc-1-2-7-rc3
libtirpc-1-2-7-rc4
libtirpc-1-2-7-rc5
libtirpc-1-3-1
libtirpc-1-3-2
libtirpc-1-3-2-rc1
libtirpc-1-3-2-rc2
libtirpc-1-3-2-rc3
libtirpc-1-3-3
libtirpc-1-3-3-rc1
libtirpc-1-3-3-rc2
libtirpc-1-3-3-rc3
libtirpc-1-3-3-rc4
libtirpc-1-3-3-rc5
libtirpc-1-3-4
libtirpc-1-3-4-rc1
libtirpc-1-3-4-rc2
libtirpc-1-3-4-rc3
libtirpc-1-3-4-rc4
libtirpc-1-3-5
libtirpc-1-3-5-rc1
libtirpc-1-3-5-rc2
libtirpc-1-3-5-rc3
libtirpc-1-3-6
libtirpc-1-3-6-rc1
libtirpc-1-3-6-rc2
libtirpc-1-3-6-rc3
libtirpc-1-3-7
libtirpc-1-3-7-rc1
libtirpc-1-3-7-rc2
libtirpc-1-3-7-rc3
libtirpc-1-3-7-rc4
```

Maybe
```suggestion
source:
  git: git://linux-nfs.org/~steved/libtirpc
  tag: libtirpc-${{ version | split(".") | join("-") }}
```

matthewfeickert replied:
@jaimergp they do, but I was hesitant to use a tag, as a tag can change. But at the same time, nothing keeps people from reuploading a tarball, so we should go with your solution. Thanks!

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,96 @@
+
+build:
+  number: 1
+  skip: not linux
```
Do we know that this doesn't work on macOS? Or is a different library used there?

cphyc replied:
libtirpc doesn't seem to compile on macOS, but I could be proven wrong!

matthewfeickert replied:
Okay, let me try, and then I'll do a review that steps through the changes. 👍

matthewfeickert commented:
@conda-forge/libtirpc this is ready for review, but as some of these changes are drastic I am fully happy to walk back anything. I just wanted to get to a spot where things were building for all platforms.

cphyc commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,96 @@
+    - cp $BUILD_PREFIX/share/gnuconfig/config.* .
+    - autoreconf --install --force
+    - ./configure --help
+    - ./configure --prefix=${PREFIX} --disable-static --disable-gssapi
```
Why the need to disable static here?

matthewfeickert replied:
Disable compiling a static library to conform to [CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md) &mdash; that is, don't package static libraries unless you need to and if you need to they need to be a `-static` suffix output.

If the static library is _needed_ (by which I mean _you_ are actively using it and can not use a dynamic library given some other library's design requirements) then the feedstock should be multi-output and follow CFEP-18.

cphyc reviewed (commented):
Thanks so much for the effort.

It is, however, somewhat challenging to review given the numerous changes (including automatically generated code). Do you have a suggestion on how to proceed?

matthewfeickert committed:
ENH: Enable macOS builds

* Enable IPV6 constants on macOS which are required to compile.
   - c.f. https://krbdev.mit.edu/rt/Ticket/History.html?id=8349

matthewfeickert committed:
TST: Add test program

* Add test program to ensure library is working as expected. The test
  program is copied from the Homebrew recipe for libtirpc.
  https://github.com/Homebrew/homebrew-core/blob/f99472a12a14f7512309ebedd2fd44072737ac34/Formula/lib/libtirpc.rb#L29-L40
* BSD 2-Clause License, Copyright (c) 2009-present, Homebrew contributors

matthewfeickert committed:
MNT: Re-rendered with conda-smithy 3.52.2 and conda-forge-pinning 2025.09.23.13.21.29

Other tools:
- conda-build 25.7.0
- rattler-build 0.47.0
- rattler-build-conda-compat 1.4.6

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+  name: ${{ name|lower }}
+  version: ${{ version }}
+
+source:
+  git: git://linux-nfs.org/~steved/libtirpc
+  tag: libtirpc-${{ version | split(".") | join("-") }}
```
As sourceforge is unreliable, get the source directly from the Git repostiory.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+  number: 1
+  skip: win
+  script:
+    - cp $BUILD_PREFIX/share/gnuconfig/config.* .
+    - autoreconf --install --force
+    - ./configure --help
```
Add support for C cross-compilation using autotools

c.f. https://conda-forge.org/docs/maintainer/knowledge_base/#cross-compilation-examples

Add `./configure --help` for visual debugging.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+    - cp $BUILD_PREFIX/share/gnuconfig/config.* .
+    - autoreconf --install --force
+    - ./configure --help
+    # Enable IPV6 constants on macOS
+    # c.f. https://krbdev.mit.edu/rt/Ticket/History.html?id=8349
+    - if: osx
+      then: export CFLAGS="${CFLAGS} -D__APPLE_USE_RFC_3542"
```
Enable macOS support by enabling IPV6 constants.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+  build:
+    - ${{ compiler('c') }}
+    - ${{ stdlib("c") }}
+    - automake
+    - libtool
+    - make
+    - pkg-config
+    - gnuconfig
```
Add autotools packages and required packages for C cross-compilation with autotools.

c.f.https://conda-forge.org/docs/maintainer/knowledge_base/#cross-compilation-examples

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+    - make
+    - pkg-config
+    - gnuconfig
+    # Tools that autotools looks for
+    - sed
+    - grep
+    - gawk
```
Add tools that autotools will look for so that as much as possible of the build tools is coming from conda-forge and not the host runner machine.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+    - grep
+    - gawk
+
+tests:
+  - package_contents:
```
Use [`rattler`'s `package_contents`](https://rattler.build/latest/testing/#testing-package-contents) to verify the contents of all files in the package. This makes the recipe a manifest as well.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+        - tirpc/rpcsvc/crypt.x
+      lib:
+        - tirpc
+  - files:
+      recipe:
+        - test.c
+    requirements:
+      run:
+        - ${{ compiler('c') }}
+    script:
+      - $CC test.c -o test $CFLAGS -I${CONDA_PREFIX}/include/tirpc $LDFLAGS -ltirpc
+      - ./test
```
Add a test program so that the library is validated to work.

matthewfeickert commented on recipe/test.c:
```diff
@@ -0,0 +1,13 @@
+#include <rpc/rpc.h>
+#include <rpc/xdr.h>
+#include <stdio.h>
+
+int main() {
+  XDR xdr;
+  char buf[256];
+  xdrmem_create(&xdr, buf, sizeof(buf), XDR_ENCODE);
+  int i = 42;
+  xdr_destroy(&xdr);
+  printf("xdr_int succeeded");
+  return 0;
+}
```
This is copied directly from the Homebrew recipe for `libtirpc`

https://github.com/Homebrew/homebrew-core/blob/f99472a12a14f7512309ebedd2fd44072737ac34/Formula/lib/libtirpc.rb#L29-L40

jaimergp replied:
This probably deserves a comment in the source file for provenance and attribution (copyright, license)

matthewfeickert replied:
@jaimergp Yeah, that's a good idea. Done and rebased.

matthewfeickert commented on conda-forge.yml:
```diff
@@ -1,6 +1,16 @@
+build_platform:
+  linux_aarch64: linux_64
+  linux_ppc64le: linux_64
+  osx_arm64: osx_64
```
Enable cross-compilation

matthewfeickert commented on conda-forge.yml:
```diff
@@ -1,6 +1,16 @@
   tooling_branch_name: main
 conda_build:
   error_overlinking: true
+conda_build_tool: rattler-build
+conda_install_tool: pixi
```
Use `rattler-build` and `pixi` as the v1 recipe is being used for substantially faster builds.

matthewfeickert reviewed (commented):
@cphyc I've updated the PR body text with a high level summary and then stepped through and commented all of the things that were changed in the PR.

Everything else is an automated change from conversion or rerendering.

I've also split the PR up into multiple commits, so if you want to you can also look at each one to get an atomic view of the process.

matthewfeickert commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+    # c.f. https://krbdev.mit.edu/rt/Ticket/History.html?id=8349
+    - if: osx
+      then: export CFLAGS="${CFLAGS} -D__APPLE_USE_RFC_3542"
+    - ./configure --prefix=${PREFIX} --disable-static --disable-gssapi
```
Why is `--disable-gssapi` enabled? 

```
--disable-gssapi        Disable GSSAPI support  [default=no]
```

cphyc reviewed (approved):
Thanks VERY much for the detailed comments.
If @jaimergp is also happy, I will merge as is and we can open another PR if support for `gssapi` is possible and useful.

jaimergp commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,110 @@
+    entries that used to be in glibc.
+  license: BSD-3-Clause
+  license_file: COPYING
+  homepage: http://git.linux-nfs.org/?p=steved/libtirpc.git;a=summary
```
Is this URL still correct? I can't get it to load.

matthewfeickert replied:
I can't get it to load either, but the URL is unchanged from the v0 recipe. I don't know what the web viewable form of the Git server is, if there is one.

jaimergp replied:
Let's open an issue for that and move on.

matthewfeickert replied:
Issue https://github.com/conda-forge/libtirpc-feedstock/issues/7
